src/main/java/seedu/todo/controllers/RedoController.java

```diff
@@ -1,26 +1,61 @@
 package seedu.todo.controllers;
 
+import java.util.HashMap;
+import java.util.Map;
+
+import seedu.todo.commons.exceptions.UnmatchedQuotesException;
+import seedu.todo.commons.util.StringUtil;
 import seedu.todo.models.TodoListDB;
 import seedu.todo.ui.UiManager;
 import seedu.todo.ui.views.IndexView;
 
 public class RedoController implements Controller {
     
-    private static final String MESSAGE_SUCCESS = "Successfully redid last command!\nTo undo, type \"undo\".";
+    private static final String MESSAGE_SUCCESS = "Successfully redid %s %s!\nTo undo, type \"undo\".";
+    private static final String MESSAGE_MULTIPLE_FAILURE = "We cannot redo %s %s! At most, you can redo %s %s.";
     private static final String MESSAGE_FAILURE = "There is no command to redo!";
 
     @Override
     public float inputConfidence(String input) {
         return input.startsWith("redo") ? 1 : 0;
     }
+    
+    private static Map<String, String[]> getTokenDefinitions() {
+        Map<String, String[]> tokenDefinitions = new HashMap<String, String[]>();
+        tokenDefinitions.put("default", new String[] {"redo"});
+        return tokenDefinitions;
+    }
 
     @Override
     public void process(String input) {
+        
+        Map<String, String[]> parsedResult;
+        try {
+            parsedResult = Tokenizer.tokenize(getTokenDefinitions(), input);            
+        } catch (UnmatchedQuotesException e) {
+            System.out.println("Unmatched quote!");
+            return;
+        }
+        
+        int numRedo = 1;
+        if (parsedResult.get("default")[1] != null) {
+            numRedo = Integer.parseInt(parsedResult.get("default")[1]);
+        }
+        
+        // We don't really have a nice way to support SQL transactions, so yeah >_<
         TodoListDB db = TodoListDB.getInstance();
-        if (!db.redo()) {
-            UiManager.updateConsoleMessage(MESSAGE_FAILURE);
+        if (db.redoSize() < numRedo || numRedo < 0) {
+            UiManager.updateConsoleMessage(String.format(MESSAGE_MULTIPLE_FAILURE,
+                    numRedo, StringUtil.pluralizer(numRedo, "command", "commands"),
+                    db.redoSize(), StringUtil.pluralizer(db.redoSize(), "command", "commands")));
             return;
         }
+        for (int i = 0; i < numRedo; i++) {
+            if (!db.redo()) {
+                UiManager.updateConsoleMessage(MESSAGE_FAILURE);
```

Comment:
Don't think this is ever reached? `MESSAGE_MULTIPLE_FAILURE` is displayed instead when there are no commands to undo. Not really an issue, but the error becomes "We cannot redo 1 command! At most, you can redo 0 commands."